dbms/src/Functions/geohashCoverAreaWithBoxes.cpp

```diff
@@ -0,0 +1,172 @@
+
+            const auto estimated_array_size = GeoUtils::geohashEstimateCoverAreaWithBoxesItems(
+                        lon_min_value, lat_min_value, lon_max_value, lat_max_value, precision_value);
+
+            if (estimated_array_size > max_array_size)
+            {
+                throw Exception{getName() + " would produce " + std::to_string(estimated_array_size) +
+                    " array elements, which is greater than the allowed maximum of " + std::to_string(max_array_size),
+                    ErrorCodes::ARGUMENT_OUT_OF_BOUND};
+            }
+
+            res_strings_offsets.reserve(res_strings_offsets.size() + estimated_array_size);
+            res_strings_chars.resize(res_strings_chars.size() + estimated_array_size * (precision_value + 1));
+            const auto starting_offset = res_strings_offsets.empty() ? 0 : res_strings_offsets.back();
+            char * out = reinterpret_cast<char *>(res_strings_chars.data() + starting_offset);
+
+            const auto actual_array_size = GeoUtils::geohashCoverAreaWithBoxes(
+                        lon_min_value, lat_min_value, lon_max_value, lat_max_value, precision_value, out);
+
+            for (UInt8 i = 1; i <= actual_array_size ; ++i)
+            {
+                res_strings_offsets.push_back(starting_offset + (precision_value + 1) * i);
+            }
+
+            res_offsets.push_back((res_offsets.empty() ? 0 : res_offsets.back()) + actual_array_size);
+        }
+        if (!res_strings_offsets.empty() && res_strings_offsets.back() != res_strings_chars.size())
+        {
+            throw Exception("String column size mismatch (internal logical error)", ErrorCodes::LOGICAL_ERROR);
+        }
+
+        if (!res_offsets.empty() && res_offsets.back() != res_strings.size())
+        {
+            throw Exception("Arrary column size mismatch (internal logical error)"
+                            + std::to_string(res_offsets.back()) + " != " + std::to_string(res_strings.size())
+                            , ErrorCodes::LOGICAL_ERROR);
+        }
+
+        result = std::move(col_res);
+        return true;
+    }
+
+    void executeImpl(Block & block, const ColumnNumbers & arguments, size_t result, size_t /*input_rows_count*/) override
+    {
+        const IColumn * lon_min = block.getByPosition(arguments[0]).column.get();
+        const IColumn * lat_min = block.getByPosition(arguments[1]).column.get();
+        const IColumn * lon_max = block.getByPosition(arguments[2]).column.get();
+        const IColumn * lat_max = block.getByPosition(arguments[3]).column.get();
+        const IColumn * prec =    block.getByPosition(arguments[4]).column.get();
+        ColumnPtr & res = block.getByPosition(result).column;
+
+        if (tryExecute<Float32, UInt8>(lon_min, lat_min, lon_max, lat_max, prec, res) ||
+            tryExecute<Float64, UInt8>(lon_min, lat_min, lon_max, lat_max, prec, res))
+            return;
```

Comment:
Looks better with two template parameters. 
Now, if the second one is the same, we probably don't need it as well.

Another thing it that this checking logic just looks strange. First you check the argument types in getReturnTypeImpl.
Then, you try to execute it one way, if it fails -- another way, then finally error out again. Why is this necessary? You know the column type, so you can just call the right function. If tryExecute encounters a wrong type, it is a logical error. So basically this entire function should be replaced with `if (column is float32) execute<Float32>() else execute<Float64>()`.

Reply:
It looks like both approaches (existing and proposed) already exist in codebase, so it is not quite clear which one is more preferable. Take a look at:
* FunctionNumericPredicate
* FunctionBitTestMany
* FunctionToStringCutToZero
* FunctionHex

And many others, for this we won't win anything in terms of simplicity nor amount of generated code, and win only a little in terms of performance (constant number of calls to `checkAndGetColumn<T>`) here is an example at call site:

proposed
```
        auto * precision = checkAndGetColumn<ColumnVector<PrecisionType>>(precision_column);

        if (const auto * lon_min = checkAndGetColumn<ColumnVector<Float64>>(lon_min_column))
        {
	        const auto * lat_min = checkAndGetColumn<ColumnVector<Float64>>(lat_min_column);
	        const auto * lon_max = checkAndGetColumn<ColumnVector<Float64>>(lon_max_column);
	        const auto * lat_max = checkAndGetColumn<ColumnVector<Float64>>(lat_max_column);
	        execute(lon_min, lat_min, lon_max, lon_max, precision, result);
        }
        else if (const auto * lon_min = checkAndGetColumn<ColumnVector<Float32>>(lon_min_column))
        {
	        const auto * lat_min = checkAndGetColumn<ColumnVector<Float32>>(lat_min_column);
	        const auto * lon_max = checkAndGetColumn<ColumnVector<Float32>>(lon_max_column);
	        const auto * lat_max = checkAndGetColumn<ColumnVector<Float32>>(lat_max_column);
	        execute(lon_min, lat_min, lon_max, lon_max, precision, result);
        }
```

vs

current
```
        if (tryExecute<Float32, UInt8>(lon_min, lat_min, lon_max, lat_max, prec, res) ||
            tryExecute<Float64, UInt8>(lon_min, lat_min, lon_max, lat_max, prec, res))
            return;
```

Reply:
Ok, both approaches exist, so how do we choose one or another? If checking the arguments twice is not necessary, we shouldn't do it. Or is it necessary, i.e. is there a case when we check the argument types in `getReturnTypeImpl`, but still get wrong types in `executeImpl`? What is this case, should we add it to the tests and mention it in the comments? Without any comments, the reason for double checking is not obvious. If the double checking is really required, this is a problem of Function interface and not of this PR, but let's make this clear.

To make the code shorter, we can write: 
`if (lat_min_column.isFloat32()) execute<Float32>(...) else execute<Float64>(...).` The execute() function should throw logical error if it encounters a column of a wrong type.

Reply:
Fixed